datasetinsights/io/gcs.py

```diff
@@ -29,16 +61,91 @@ def upload(self, localfile, bucket_name, object_key):
 
         blob.upload_from_filename(localfile)
 
+    def _download_folder(self, bucket, key, local_path):
+        """ download all files from directory
+        """
+        blobs = bucket.list_blobs(prefix=key)
+        for blob in blobs:
+            local_file_path = blob.name.replace(key, local_path)
+            self._download_validate(blob, key, local_file_path)
 
-def gcs_bucket_and_path(url):
+    def _download_file(self, bucket, key, local_path):
+        """ download single file
+        """
+        blob = bucket.get_blob(key)
+        key_suffix = key.replace("/" + basename(key), "")
+        local_file_path = blob.name.replace(key_suffix, local_path)
+        self._download_validate(blob, key, local_file_path)
+
+    def _download_validate(self, blob, key, local_file_path):
+        """ download file and validate checksum
+        """
+        dst_dir = local_file_path.replace("/" + basename(local_file_path), "")
+        self._download_blob(blob, dst_dir, key, local_file_path)
+        self._checksum(blob, local_file_path)
+
+    def _download_blob(self, blob, dst_dir, key, local_file_path):
```

Comment:
What is the parameter key used for? It seems the information needed is in the blob object already.